feat: expose extrinsics operations as a library

tareknaser360 opened the pull request:
## Description

This pull request introduces a library that exposes extrinsics operations as a reusable component. This makes it easier to integrate and reuse these operations in various projects.

Fixes #1140


## Notes to Reviewers:
- Most of the changes in this PR involve reorganizing existing code. No new code has been added.
- The `runtime_api` module has been moved to the new crate to avoid cyclic dependencies.
- The types `CodeHash`, `Balance`, and  `Client` have been defined in the new crate for the same reason.
- The documentation for the extrinsics has been moved to the readme of the crate itself.

ascjones commented on crates/extrinsics/src/lib.rs:
```diff
@@ -20,6 +20,7 @@ mod error;
 mod events;
 mod instantiate;
 mod remove;
+pub mod runtime_api;
```
I'm not sure about exposing this entire module, there's a lot in there and it can easily cause breaking changes any time we update the metadata. We could

1. Move the code which fetches the `ContractInfo` into the `extrinsics` crate and export that.
2. Only export `ContractInfo` and the function used to fetch it.

What do you think?

tareknaser360 replied:
`fetch_contract_info` takes `InfoCommand` as a parameter. 
The `extrinsics` crate doesn't have access to it.
As far as I am concerned, we can't use it without having cyclic dependencies.

Is there another way around this?

ascjones replied:
It only uses `&self.contract` from `InfoCommand` so could just pass that as a parameter instead.

For the return type could even hide the `ContractInfo` type and return `InfoToJson` so we control the API instead of the codegen.

tareknaser360 replied:
> It only uses &self.contract from InfoCommand so could just pass that as a parameter instead.

Yes that's fine

but for the return type. `InfoToJson` is defined in `cargo_contract` crate so it can't be the return type of the new function in the `extrinsics` crate.
and even if we have it, how can we cast `ContractInfo` to `InfoToJson`?

ascjones replied:
Move `InfoToJson` to `extrinsics`, maybe just rename it to `ContractInfo`.

> and even if we have it, how can we cast ContractInfo to InfoToJson?

Just build the instance as we already do:

```rust
                    let convert_trie_id = hex::encode(info_result.trie_id.0);
                    let info_to_json = InfoToJson {
                        trie_id: convert_trie_id,
                        code_hash: info_result.code_hash,
                        storage_items: info_result.storage_items,
                        storage_item_deposit: info_result.storage_item_deposit,
                    };
```

tareknaser360 replied:
Thanks for the detailed answer!
Let me know if this is now resolved.

ascjones commented on crates/extrinsics/LICENSE:
This should be symbolic link to `../../LICENSE`

ascjones commented on crates/extrinsics/src/call.rs:
```diff
@@ -16,33 +16,27 @@
-    },
-    DEFAULT_KEY_COL_WIDTH,
-};
+use crate::DEFAULT_KEY_COL_WIDTH;
```
Could be moved inside `super` as with others.

ascjones commented on crates/extrinsics/src/instantiate.rs:
```diff
@@ -16,33 +16,27 @@
-    },
-    DEFAULT_KEY_COL_WIDTH,
-};
+use crate::DEFAULT_KEY_COL_WIDTH;
```
Same here, could go inside the `super`.

ascjones reviewed (approved):
LGTM. Thanks!

tareknaser360 commented:
> Hey @tareknaser360, some of the tests e.g. for `transcoder` are failing, which is why the CI is red. Can you resolve those issues then we can merge this.

I looked into the failing tests in the CI environment and made the following updates:
- Updated the examples in the new crates' documentation to resolve the test failures. 
- Modified the examples to initialize `BalanceVariant` instances using strings, which is the most suitable method at the moment.

I also rebased on top of master.
